lib/bookmark_query.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+##
+# Allows us to query Bookmark records for lists. Used mainly
+# in the user/activity/bookmarks page.
+
+class BookmarkQuery
+  def initialize(user, params)
+    @user = user
+    @params = params
+  end
+
+  def list_all
+    results = user_bookmarks
+      .joins('INNER JOIN topics ON topics.id = bookmarks.topic_id')
+      .joins('INNER JOIN posts ON posts.id = bookmarks.post_id')
+      .joins('INNER JOIN users ON users.id = posts.user_id')
+      .select(<<-SQL
```

Comment:
`select` in a case like this is quite dangerous. If in the future someone adds a column to these tables and doesn't update this code, another user of this method will have objects that look like `Bookmark` but are missing fields.

It is much safer to return the default ActiveRecord objects. Is there a specific reason you chose to do it this way?

Reply:
I was just trying to be more efficient here by selecting only the columns I need. If it is a premature optimization I will just change this to include the user/post/topic instead!

Reply:
If you want to select only certain columns you should use `pluck` or write a custom SQL query using `DB.exec`

`select` is dangerous. It returns objects that look like ActiveRecord objects but are missing certain attributes. I wrote about it here:

https://eviltrout.com/2013/10/05/enemy-of-the-state.html.html

In this case I do think it's a premature optimization and probably not worth the effort of migrating to mini-sql or pluck, but if you took those approaches I'd feel much safer. I consider `select` a bug waiting to happen in the future.

Reply:
Thank you for the blog article, I never thought about it in that way. Probably because I have never worked with many other people so everyone else usually knows all the columns :sweat_smile: I will change it to a regular query.